service/src/main/java/bio/terra/tanagra/plugin/Plugin.java

```diff
@@ -0,0 +1,7 @@
+package bio.terra.tanagra.plugin;
+
+import javax.sql.DataSource;
+
+public interface Plugin {
+  void init(PluginConfig config, DataSource dataSource);
```

Comment:
Plugins should not use the main application Spring data sources. For plugins that need to store some internal state, here are 3 possible ways to do that:

1. Maintain its own DB as part of a companion service. Define an access control DB schema and have VUMC-specific fields. Call the companion service from an access control plugin in core codebase.
  - In this scenario, the companion service would have it's own Spring configuration file and you'd define the datasource connection/etc. info there. In the short term, if you want to code in the core Tanagra repo with plans to pull out the companion service into a separate repo later, I'd just add the datasource you want directly to the Spring yaml and we can just pull that part out when we refactor later. There's no need to make this configuration generic (i.e. plugin name + data source info) here because there should only be one other datasource and it would be temporary just for this initial implementation.

2. Use core Tanagra's existing DB (where cohorts/studies/etc are stored). Define an access control DB schema with no VUMC-specific fields. Query this DB from an access control plugin in core codebase. This is great in that it adds functionality to core Tanagra, but seems like potentially more work upfront to define and implement a generalizable access control model. Ideally we'd have product info on other VUMC datasets before doing this, but fine to make an educated first guess if you'd like.
  - In this scenario, you'd just use the existing data source info (no need to define another) and modify the existing schema. You can look at the `*Dao` classes for example code here.

3. Use core Tanagra's study metadata to store access control information. No DB schema to define, it's just a string-only key-value map. You can store whatever you want in there (e.g. list of emails of users who can access the study, serialized representation of some object), and then pull it out to determine access. This is sort of messy and less maintainable, but probably quicker to get working.
  - In this scenario, you wouldn't need to access the datasource info directly at all. Instead, you'd just save the study (e.g. by calling methods in `StudyService`) and piggyback on the existing db read/write code.

Reply:
I disagree. There is no reason that Tanagra core can't provide raw data sources to a plugin, the alternative is basically copy/pasting a bunch of boilerplate code into individual plugins. I agree that it shouldn't be on Tanagra to manage these resources, but instantiating them based on configuration and handing them of to a plugin should be in scope.

Reply:
My thinking of how not-built-in plugins (e.g. a VUMC-specific one) should work is:
- The code should live in a separate repo. There is an implementation of at least one interface that is defined in core Tanagra (e.g. `AccessControlPlugin`). This repo outputs a JAR file.
- The VUMC underlay config file specifies the implementation class name (e.g. `SddAccessControlPlugin`).
- Core Tanagra is deployed with the other-repo JAR file on the classpath.
- At runtime, core Tanagra uses reflection to instantiate the implementation class from the JAR on its classpath.

My reasoning here is that plugins should not touch core Tanagra code, only config files. That means they don't modify the Spring application yaml to specify a custom data source.

So if a plugin wants to use a Spring data source, then I think it should either:
- Live in a companion service, where it can define that data source in its own application yaml.
- Be a built-in plugin (i.e. one that we package with core Tanagra) that uses core Tanagra's own DB. In that case, it can just use the data source already defined in core Tanagra's application yaml.

Reply:
The VUMC-specific code should ultimately live in a separate resource yes, but it still has to be built against core Tanagra in order to import core Tanagra objects such as the plugin interface, `UserId`, `ResourceId`, and `ResourceType`. Why is passing a data source to a plugin, which is configured in an underlay loaded by the core any different?

Reply:
I think the reason we want a plugin architecture is so that you can override core Tanagra functionality without touching core Tanagra code, only underlay config files. Otherwise, we could just write everything into core Tanagra and use feature flags (this is option 2, and also a fine approach, but then we can just use the single data source). So if you have to modify core Tanagra's Spring configuration in order for your plugin to work, then we haven't got the division quite right.

I also think that the Spring data source should be defined in the same place as the DB schema and associated upgrade/migration utilities, which I expect would be in the separate repo along with the code that relies on it.

Reply:
Alright.